File: Backend/models/Incident.js
const mongoose = require("mongoose");

// ─── Media sub-document ──────────────────────────────────────────────────────
const mediaSchema = new mongoose.Schema(
    {
        url: {
            type: String,
            required: true,
        },
        filename: {
            type: String,
            default: "",
        },
        provider: {
            type: String,
            enum: ["local", "cloudinary"],
            default: "local",
        },
        publicId: {
            type: String,
            default: null,
        },
        resourceType: {
            // "image" | "video" as returned by Cloudinary
            type: String,
            default: "image",
        },
        capturedAt: {
            // EXIF capture time from mobile, falls back to upload time
            type: Date,
            default: null,
        },
    },
    { _id: false }
);

// ─── Incident schema ──────────────────────────────────────────────────────────
const incidentSchema = new mongoose.Schema(
    {
        // ── Who reported ──────────────────────────────────────────────────────
        reportedBy: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true,
            index: true,
        },

        // ── Classification ────────────────────────────────────────────────────
        incidentType: {
            type: String,
            required: true,
            enum: [
                "landslide_crack",
                "blocked_road",
                "slope_movement",
                "soil_erosion",
                "flooding",
                "bridge_damage",
                "other",
            ],
            index: true,
        },

        severity: {
            type: String,
            required: true,
            enum: ["low", "medium", "high", "critical"],
            default: "medium",
            index: true,
        },

        description: {
            type: String,
            required: true,
            trim: true,
            maxlength: 3000,
        },

        // ── GeoJSON location (2dsphere indexed for spatial queries) ───────────
        location: {
            type: {
                type: String,
                enum: ["Point"],
                default: "Point",
            },
            coordinates: {
                // [longitude, latitude] — GeoJSON standard
                type: [Number],
                required: true,
                validate: {
                    validator: (coords) =>
                        Array.isArray(coords) &&
                        coords.length === 2 &&
                        coords[0] >= -180 &&
                        coords[0] <= 180 &&
                        coords[1] >= -90 &&
                        coords[1] <= 90,
                    message:
                        "Coordinates must be [longitude, latitude] within valid ranges.",
                },
            },
        },

        // ── NE Region context metadata ─────────────────────────────────────────
        locationMeta: {
            address: { type: String, trim: true, default: "" },
            district: { type: String, trim: true, default: "", index: true },
            state: {
                type: String,
                trim: true,
                default: "",
                // NE states covered by the SIH problem statement
                enum: [
                    "Arunachal Pradesh",
                    "Assam",
                    "Manipur",
                    "Meghalaya",
                    "Mizoram",
                    "Nagaland",
                    "Sikkim",
                    "Tripura",
                    "",
                ],
            },
            altitude: {
                // metres above sea level — relevant for slope risk
                type: Number,
                default: null,
            },
            slopeAngle: {
                // degrees — key input for landslide probability
                type: Number,
                min: 0,
                max: 90,
                default: null,
            },
        },

        // ── Field observations ────────────────────────────────────────────────
        witnessCount: {
            type: Number,
            min: 0,
            default: 1,
        },

        isRoadBlocked: {
            type: Boolean,
            default: false,
        },

        affectedVillages: {
            type: [String],
            default: [],
        },

        // ── Auto-computed risk score 0–100 (set by pre-save hook) ─────────────
        riskScore: {
            type: Number,
            min: 0,
            max: 100,
            default: 0,
            index: true,
        },

        // ── Uploaded media (images + short videos) ────────────────────────────
        media: {
            type: [mediaSchema],
            default: [],
        },

        // ── Workflow status ───────────────────────────────────────────────────
        status: {
            type: String,
            enum: ["pending", "verified", "escalated", "resolved", "rejected"],
            default: "pending",
            index: true,
        },

        verifiedBy: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            default: null,
        },

        verifiedAt: {
            type: Date,
            default: null,
        },

        assignedTo: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            default: null,
        },

        remarks: {
            type: String,
            trim: true,
            default: "",
        },

        // ── Offline sync support ──────────────────────────────────────────────
        // Client-generated UUID; unique sparse index allows safe retries
        offlineId: {
            type: String,
            default: null,
            trim: true,
        },

        syncedAt: {
            type: Date,
            default: () => new Date(),
        },
    },
    {
        timestamps: true,
    }
);

// ─── Indexes ──────────────────────────────────────────────────────────────────
incidentSchema.index({ location: "2dsphere" });
incidentSchema.index({ offlineId: 1 }, { unique: true, sparse: true });
incidentSchema.index({ createdAt: -1 });
incidentSchema.index({ incidentType: 1, severity: 1, status: 1 });

// ─── Pre-save: compute riskScore ──────────────────────────────────────────────
const SEVERITY_BASE = { low: 15, medium: 35, high: 65, critical: 90 };

incidentSchema.pre("save", function (next) {
    const base = SEVERITY_BASE[this.severity] || 35;

    // Witness boost: each additional witness adds up to 5 points (cap 10)
    const witnessBoost = Math.min((this.witnessCount - 1) * 2.5, 10);

    // Road blocked adds urgency
    const roadBoost = this.isRoadBlocked ? 5 : 0;

    // Slope angle amplifier: steep slopes get up to 10 extra points
    const slopeBoost =
        this.locationMeta?.slopeAngle != null
            ? Math.min((this.locationMeta.slopeAngle / 90) * 10, 10)
            : 0;

    this.riskScore = Math.min(
        Math.round(base + witnessBoost + roadBoost + slopeBoost),
        100
    );

    next();
});

module.exports = mongoose.model("Incident", incidentSchema);

File: ml_backend/database/models/incident.py
"""
Incident SQLAlchemy model representing disaster events, ground reports, and field hazard observations.
"""

from __future__ import annotations

import enum
from datetime import datetime
from typing import TYPE_CHECKING, Any, Dict, List, Optional

from sqlalchemy import Boolean, DateTime, Enum, Float, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ml_backend.database.models.base import Base, JSONType, SoftDeleteMixin, TimestampMixin, UUIDMixin

if TYPE_CHECKING:
    from ml_backend.database.models.responder import Responder
    from ml_backend.database.models.user import User


class IncidentType(str, enum.Enum):
    """Types of disaster incidents."""

    LANDSLIDE_CRACK = "landslide_crack"
    BLOCKED_ROAD = "blocked_road"
    SLOPE_MOVEMENT = "slope_movement"
    SOIL_EROSION = "soil_erosion"
    FLOODING = "flooding"
    FLOOD = "flooding"
    BRIDGE_DAMAGE = "bridge_damage"
    CYCLONE = "cyclone"
    EARTHQUAKE = "earthquake"
    WILDFIRE = "wildfire"
    TSUNAMI = "tsunami"
    OTHER = "other"


class IncidentSeverity(str, enum.Enum):
    """Incident severity classification."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class IncidentStatus(str, enum.Enum):
    """Workflow statuses for an incident."""

    PENDING = "pending"
    VERIFIED = "verified"
    ESCALATED = "escalated"
    RESOLVED = "resolved"
    REJECTED = "rejected"


class Incident(Base, UUIDMixin, TimestampMixin, SoftDeleteMixin):
    """Incident report model."""

    __tablename__ = "incidents"

    # Reporter reference
    reported_by_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Classification
    incident_type: Mapped[IncidentType] = mapped_column(
        Enum(IncidentType, native_enum=False, length=30),
        nullable=False,
        index=True,
    )
    severity: Mapped[IncidentSeverity] = mapped_column(
        Enum(IncidentSeverity, native_enum=False, length=20),
        default=IncidentSeverity.MEDIUM,
        nullable=False,
        index=True,
    )
    title: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")

    # Geolocation (WGS-84 decimal coordinates)
    latitude: Mapped[float] = mapped_column(Float, nullable=False, index=True)
    longitude: Mapped[float] = mapped_column(Float, nullable=False, index=True)

    # Spatial context & terrain parameters
    address: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    district: Mapped[Optional[str]] = mapped_column(String(100), nullable=True, index=True)
    state: Mapped[Optional[str]] = mapped_column(String(100), nullable=True, index=True)
    altitude: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    slope_angle: Mapped[Optional[float]] = mapped_column(Float, nullable=True)

    # Field observations
    witness_count: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    is_road_blocked: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    affected_villages: Mapped[List[str]] = mapped_column(JSONType, default=list, nullable=False)
    risk_score: Mapped[float] = mapped_column(Float, default=0.0, nullable=False, index=True)

    # Media attachments (URLs, captions, metadata)
    media: Mapped[List[Dict[str, Any]]] = mapped_column(JSONType, default=list, nullable=False)

    # Workflow & verification
    status: Mapped[IncidentStatus] = mapped_column(
        Enum(IncidentStatus, native_enum=False, length=20),
        default=IncidentStatus.PENDING,
        nullable=False,
        index=True,
    )
    verified_by_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    verified_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    assigned_to_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    remarks: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Offline synchronization metadata
    offline_id: Mapped[Optional[str]] = mapped_column(String(64), nullable=True, unique=True, index=True)
    synced_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)

    # Relationships
    reported_by: Mapped[User] = relationship(
        "User",
        foreign_keys=[reported_by_id],
        back_populates="reported_incidents",
    )
    verified_by: Mapped[Optional[User]] = relationship(
        "User",
        foreign_keys=[verified_by_id],
        back_populates="verified_incidents",
    )
    assigned_to: Mapped[Optional[User]] = relationship(
        "User",
        foreign_keys=[assigned_to_id],
        back_populates="assigned_incidents",
    )
    assigned_responders: Mapped[List[Responder]] = relationship(
        "Responder",
        back_populates="active_incident",
    )

    __table_args__ = (
        Index("ix_incidents_lat_lon", "latitude", "longitude"),
        Index("ix_incidents_status_severity", "status", "severity"),
        Index("ix_incidents_state_district", "state", "district"),
    )
